### PR TITLE
Generate and write stash including soft blocked versions

### PR DESCRIPTION
**Pull request by KevinMind**
Fixes: mozilla/addons#15050

### Description

This PR updates the `generate_and _write_stash` method on `MLBF` class to consider the range of use cases covered in the linked jira issue document.

We have to add support for specifying a block_type when requesting a change count. This will be useful for creating type level filters as we need to know for each type if we have any changes and enough changes for a new filter.

We have to modify which guids are included in the now 3 different properties on the stash. This has to cover the set of use cases where blocks are added as hard or soft as well as moving from soft to hard and vice versa.

### Context

When updating our bloom filter we are now going to create a unified stash which will allow us to enable support for adding/removing items from the soon to be soft block filter, but, to do so in a way that is backwards compatible with legacy fx clients that have no concept of soft blocking. This raises some edge cases that increased the complexity of this patch. Read the doc for more info

### Testing

Look at the automated tests but essentially we should consider the following scenarios and you can independently verify them if you don't trust the tests.

Here is a [detailed break ](https://github.com/mozilla/addons-server/pull/22826/files#diff-9df60b77ee1f859628ea78fa4cbe1729ea4dc9d67d0a76c77da8003e52c1fba9R393-R448)down of the scenarios

### Checklist

- [X] Add `#ISSUENUM` at the top of your PR to an existing open issue in the mozilla/addons repository.
- [X] Successfully verified the change locally.
- [X] The change is covered by automated tests, or otherwise indicated why doing so is unnecessary/impossible.
- [ ] Add before and after screenshots (Only for changes that impact the UI).
- [ ] Add or update relevant [docs](../docs/) reflecting the changes made.

### CODE REVIEW

**Comment by KevinMind** [comment]
@willdurand and @diox 2 remaining questions from my side.

1. what logic should we use to determine the number of "changed" guids when including the differing handling of hard <-> soft blocks in legacy versus new clients. Some items are included in multiple stashes so should they equally double increase the counts? Seems intuitive enough but not sure if we care about that.
2. It is important that the time used for the stash is earlier than the time used for the filter in the event that we create both. Currently we use a single generation time for both. Do we want to always make the filter generation time +1 milisecond of the stash? or vice versa or how exactly should we handle this?

File: src/olympia/blocklist/mlbf.py [hunk 1]
```diff
@@ -240,20 +240,37 @@ def generate_diffs(
         }
 
     def generate_and_write_stash(self, previous_mlbf: 'MLBF' = None):
-        # compare previous with current blocks
-        extras, deletes, _ = self.generate_diffs(previous_mlbf)[BlockType.BLOCKED]
+        """
+        Generate and write the stash file representing
+        changes between the previous and current MLBF Filters.
+        In order to support older FX clients that only understood blocked and unblocked
+        The unblocked list is a union of deletions from blocked and soft_blocked.
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:247]: Where is this coming from?

Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:247]: Unblocked includes versions which have gone from hard blocked or from soft blocked to unblocked. It also includes versions which have gone from hard blocked to soft blocked.

Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:247]: This is attempting to implement what is written in the google doc you wrote.

Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py]: I updated the doc string to be a bit clearer. lmk

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 2]
```diff
@@ -300,19 +311,217 @@ def test_diff_block_added_and_removed(self):
...
+    def test_diff_all_possible_changes(self):
+        """
+        Simulates 6 guid:version comibinations moving from one state to another
+        covering all possible movements and the final diff state.
+
+        1. Not blocked -> Soft blocked
+        2. Not blocked -> Blocked
+        3. Soft blocked -> Blocked
+        4. Soft blocked -> Not blocked
+        5. Blocked -> Soft blocked
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:442]: Use of the word this should be forbidden in code reviews. Which this are you referring to here?

Each one of these edge cases is important to cover for either modern or legacy versions of FX and moreover we have to ensure the backward compatibility that we agreed on (that you proposed) does not break either legacy or modern Fx clients.

**Comment by KevinMind** [comment]
@willdurand rebased

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:464]: _This_ line 😬  Not sure if that makes the code more complex or anything, but there won't be hard to soft block transitions.

File: src/olympia/blocklist/mlbf.py [hunk 3]
```diff
@@ -245,17 +245,42 @@ def generate_diffs(
         }
 
     def generate_and_write_stash(self, previous_mlbf: 'MLBF' = None):
-        # compare previous with current blocks
-        extras, deletes, _ = self.generate_diffs(previous_mlbf)[BlockType.BLOCKED]
+        """
+        Generate and write the stash file representing
+        changes between the previous and current MLBF Filters.
+
+        In order to support older FX clients that only understood blocked and unblocked
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:252]:
```suggestion
        In order to support older FX clients that only understood blocked and unblocked,
```

File: src/olympia/blocklist/mlbf.py [hunk 4]
```diff
@@ -245,17 +245,42 @@ def generate_diffs(
...
+        unblocked is a union of deletions from blocked and additions to soft_blocked.
+
+        Versions that move from hard to soft blocked will be picked up by old clients
+        as no longer hard blocked by being in the unblocked list.
+
+        Newer clients will also see soft blocked items as unblocked
+        but then will check the soft blocked list and re-add them to soft blocked.
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:259]: I am reading the PR top to bottom and this is one of the first comments... Q: is this adding complexity in the patch? If so, we can probably do without it because Product + others said we didn't need to account for hard to soft transitions.

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 5]
```diff
@@ -452,6 +452,145 @@ def test_diff_block_soft_to_hard(self):
...
+        6. Blocked -> Not blocked
+
+        first stash
+        {
+          'blocked': [5,6],
+          'softblocked': [3,4],
+          # 1,2 are already unblocked so wouldn't be on the stash
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:471]: I am not clear on why that's `true`.

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 6]
```diff
@@ -452,6 +452,145 @@ def test_diff_block_soft_to_hard(self):
...
+        (five_hash,) = MLBF.hash_filter_inputs(
+            [(five.guid, five.current_version.version)]
+        )
+        # 6. Blocked -> Not blocked
+        six, six_block = self._blocked_addon(
+            guid='6', block_type=BlockType.BLOCKED, file_kw={'is_signed': True}
+        )
+        (six_hash,) = MLBF.hash_filter_inputs([(six.guid, six.current_version.version)])
+
+        first_mlbf = MLBF.generate_from_db('first')
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:547]: I am not clear on why the first mlbf would generate differences if we create all the blocks _before_ we generate the bloom filter from the DB.

**Review by willdurand — COMMENTED** [review]
I have questions.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:464]: There is a PR introducing that feature (hard->soft) now so AFAIK yes it will be possible to do. Even if we don't use hard->soft feature I still think it makes sense to account for the total possible transitions, MLBF shouldn't care about how we intend to use the feature it just works.. esp since the code covering this case is already here and tested.

hard to soft results in a version being in both unblocked and soft-blocked on the next stash

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:471]: What what is true? 1 and 2 were blocked before we re-generated the stash so we don't add them to "unblocked".

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 6 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:547]: If there is no previous filter to compare to, then the filter will consider everything in the DB a "change" because, well, it is. Nothing to something.

T0: we create all the blocks
T1: we create the first_mlbf (it includes all of the "changes" from T0. since there is no previous filter to compare to everything is a change)
T2: we create and update some records
T3: we create another filter and now track the changes from T1 to T3

Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:547]: FWIW that is existing behavior and is how regenerating the filter works.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:464]: Ok, that's fine.

File: src/olympia/blocklist/mlbf.py (hunk 4 above)
Comment (willdurand) [reply src/olympia/blocklist/mlbf.py:259]: Well, it might not be the right question but that is still _my_ question, and a question that I want an answer for. That said, the code doesn't look too complicated so probably fine.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:471]: Comments in GitHub diffs are contextual. They apply to individual lines (or multiple lines, sometimes).

> 1 and 2 were blocked before we re-generated the stash so we don't add them to "unblocked".

That wasn't entirely clear to me. I read the comment then jumped to the generation of the MLBF and the assertion, and didn't quite get how we ended in this state. Maybe splitting the long python comment into inlined comments in the test would be easier to follow? I don't know.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 6 above)
Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:547]: OK so it looks like we don't create all the blocks for all the versions there, which is what I assumed. The first two versions are not blocked at this point. We block them after the initial MLBF is generated.

File: src/olympia/blocklist/mlbf.py (hunk 4 above)
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:259]: I was trying to answer your question.. hope that I did. To quote the python style guide "simple is better than complex" but "complex is better than complicated" I think this patch is somewhere between simple and complex, but at least not complicated. All the complexity is in one place and changes elsewhere will not impact or be impacted by this code. Closing this thread.

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 6 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:547]: Correct, following the spec in the test description. I think you should now understand [why 1,2 were already unblocked](https://github.com/mozilla/addons-server/pull/22826#discussion_r1837935103). Question is, how could I update the test/comments to make it more clear from the start?

File: src/olympia/blocklist/tests/test_mlbf.py (hunk 5 above)
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:471]: I know which line you referred to but it wasn't clear to me exactly which part of the statement being true is not clear. Would something like

"Versions that were already not hard or softblocked will not be included in any stash since their state has not changed"

Would that have helped do you think?

Comment (willdurand) [reply src/olympia/blocklist/tests/test_mlbf.py:471]: Not really. I think it's more mapping parts of the large comment to what happens in the test that could be helpful.

Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py]: Updated by removing the long comment and trying to make the inline comments more clear.

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 7]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        )
+        # 6. (Blocked) -> (Not blocked)
+        six, six_block = self._blocked_addon(
+            guid='6', block_type=BlockType.BLOCKED, file_kw={'is_signed': True}
+        )
+        (six_hash,) = MLBF.hash_filter_inputs([(six.guid, six.current_version.version)])
+
+        first_mlbf = MLBF.generate_from_db('first')
+
+        assert first_mlbf.generate_diffs() == {
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:505]:
```suggestion
        # We expect the 4 blocked versions to be in the diff, sorted by block type.
        assert first_mlbf.generate_diffs() == {
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 8]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+            BlockType.BLOCKED: ([five_hash, six_hash], [], 2),
+            BlockType.SOFT_BLOCKED: ([three_hash, four_hash], [], 2),
+        }
+
+        # The first time we generate the stash, we expect 3-6 to be in the stash
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:510]:
```suggestion
        # The first time we generate the stash, we expect 3 to 6 to be in the stash
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 9]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        # as they have some kind of block applied
+        assert first_mlbf.generate_and_write_stash() == {
+            'blocked': [five_hash, six_hash],
+            'softblocked': [three_hash, four_hash],
+            'unblocked': [],
+        }
+
+        # Transform the addons to the next state
+
+        # 1. Not blocked -> (Soft blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:526]:
```suggestion
        # The first version gets soft blocked now.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 10]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        block_factory(
+            guid=one.guid, updated_by=self.user, block_type=BlockType.SOFT_BLOCKED
+        )
+        # 2. Not blocked -> (Blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:530]:
```suggestion
        # The second version is hard blocked.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 11]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        block_factory(guid=two.guid, updated_by=self.user, block_type=BlockType.BLOCKED)
+        # 3. Soft blocked -> (Blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:532]:
```suggestion
        # 3 was soft-blocked and is now hard blocked.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 12]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        three_block.blockversion_set.first().update(block_type=BlockType.BLOCKED)
+        # 4. Soft blocked -> (Not blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:534]:
```suggestion
        # 4 was soft blocked and is now unblocked.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 13]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        four_block.delete()
+        # 5. Blocked -> (Soft blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:536]:
```suggestion
        # 5 was hard blocked and is now soft blocked.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 14]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        five_block.blockversion_set.first().update(block_type=BlockType.SOFT_BLOCKED)
+        # 6. Blocked -> (Not blocked)
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:538]:
```suggestion
        # 6 was hard blocked and is now unblocked.
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 15]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        six_block.delete()
+
+        second_mlbf = MLBF.generate_from_db('second')
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:541]:
```suggestion

        # We regenerate another MLBF based on the updates we've just done to verify the final
        # state of each version.
        second_mlbf = MLBF.generate_from_db('second')
```

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 16]
```diff
@@ -452,6 +452,109 @@ def test_diff_block_soft_to_hard(self):
...
+        # The order is based on the ID (i.e. creation time) of the block,
+        # not the version so we expect two after three
+        # since two was blocked after three
+        assert second_mlbf.generate_diffs(previous_mlbf=first_mlbf) == {
+            BlockType.BLOCKED: (
+                [three_hash, two_hash],
+                [five_hash, six_hash],
+                4,
+            ),
+            # Same as above, one had a block created after five so it comes second
```
Comment (willdurand) [inline src/olympia/blocklist/tests/test_mlbf.py:552]: nit

```suggestion
            # Same as above, one had a block created after five so it comes second.
```

**Review by willdurand — CHANGES_REQUESTED** [review]
This looks good overall but given the number of suggestions, I'm going to request changes to get a chance to re-read the entire diff once updated. Given that most of the tests are in the single `test_diff_all_possible_changes` test method, I made numerous suggestions to get it better documented for our future selves (not that it wasn't well documented initially).

Another thing that I am wondering is that we do a lot of comparisons between a MLBF (so a base bloom filter) and changes. Do we have test coverage to generate a stash after a previous stash was created?

File: src/olympia/blocklist/mlbf.py [hunk 17]
```diff
@@ -245,17 +245,42 @@ def generate_diffs(
...
+        Generate and write the stash file representing
+        changes between the previous and current MLBF Filters.
+
+        In order to support older FX clients that only understood blocked and unblocked,
+        unblocked is a union of deletions from blocked and additions to soft_blocked.
+
+        Versions that move from hard to soft blocked will be picked up by old clients
+        as no longer hard blocked by being in the unblocked list.
+
+        Newer clients will also see soft blocked items as unblocked
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py]: Done

File: src/olympia/blocklist/mlbf.py [hunk 18]
```diff
@@ -245,17 +245,42 @@ def generate_diffs(
...
+        but then will check the soft blocked list and re-add them to soft blocked.
+
+        Versions that move from soft to hard blocked will be picked up by old clients
+        and new clients in the blocked list. However, we prevent these versions from
+        also being in the unblocked list as this would produce a contradiction
+        since unblocked is shared by hard and soft blocked filters.
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py]: It's exactly that scenario that I was trying to describe. Updated the comment to be more explicit.

File: src/olympia/blocklist/mlbf.py [hunk 19]
```diff
@@ -245,17 +245,42 @@ def generate_diffs(
...
+        blocked_added, blocked_removed, _ = diffs[BlockType.BLOCKED]
+        soft_blocked_added, soft_blocked_removed, _ = diffs[BlockType.SOFT_BLOCKED]
         stash_json = {
-            'blocked': extras,
-            'unblocked': deletes,
+            'blocked': blocked_added,
+            'softblocked': soft_blocked_added,
+            'unblocked': [
+                unblocked
+                for unblocked in (blocked_removed + soft_blocked_removed)
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py]: Fixed

**Comment by KevinMind** [comment]
@willdurand I'm not sure I get it but

> Another thing that I am wondering is that we do a lot of comparisons between a MLBF (so a base bloom filter) and changes. Do we have test coverage to generate a stash after a previous stash was created?

Each time we run the cron job we check the current state of the database from the previous `cache.json` and if there are >0 < threshold number of changes we create a stash.

It's actually totally irrelevant if we created a stash the last time or not. It only matters what was the state of the DB during the last run. In the case where we don't have any filter to compare to, we cannot create a stash because we need to create a filter.

**Comment by willdurand** [comment]
> @willdurand I'm not sure I get it but
> 
> > Another thing that I am wondering is that we do a lot of comparisons between a MLBF (so a base bloom filter) and changes. Do we have test coverage to generate a stash after a previous stash was created?
> 
> Each time we run the cron job we check the current state of the database from the previous `cache.json` and if there are >0 < threshold number of changes we create a stash.
> 
> It's actually totally irrelevant if we created a stash the last time or not. It only matters what was the state of the DB during the last run. In the case where we don't have any filter to compare to, we cannot create a stash because we need to create a filter.

You got it right, I think. Do we have test coverage that reads a previous `cache.json` file? Something a bit more end-to-end than calling individual methods.

File: src/olympia/blocklist/mlbf.py [hunk 20]
```diff
@@ -245,17 +246,52 @@ def generate_diffs(
...
+        diffs = self.generate_diffs(previous_mlbf)
+        blocked_added, blocked_removed, _ = diffs[BlockType.BLOCKED]
         stash_json = {
-            'blocked': extras,
-            'unblocked': deletes,
+            'blocked': blocked_added,
+            'unblocked': blocked_removed,
         }
+
+        if waffle.switch_is_active('mlbf-soft-blocks-enabled'):
```
Comment (willdurand) [inline src/olympia/blocklist/mlbf.py:277]: Ha, I see. So there is the waffle switch. Given that the migration file is short, I would just add it in this patch, rather than in its own PR without anything else.

Comment (KevinMind) [reply src/olympia/blocklist/mlbf.py:277]: It's referenced in both PRs. Why not just land it?

**Review by willdurand — APPROVED** [review]
r+wc, thank you!

File: src/olympia/blocklist/tests/test_mlbf.py [hunk 21]
```diff
@@ -452,6 +454,129 @@ def test_diff_block_soft_to_hard(self):
...
+            ),
+            # Same as above, one had a block created after five so it comes second.
+            BlockType.SOFT_BLOCKED: (
+                [five_hash, one_hash],
+                [three_hash, four_hash],
+                4,
+            ),
+        }
+
+        assert second_mlbf.generate_and_write_stash(previous_mlbf=first_mlbf) == {
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply src/olympia/blocklist/tests/test_mlbf.py:567]: I think we will remove the switch before flipping the default, and even if we did, I'd just update the tests then.

**Comment by KevinMind** [comment]
@willdurand added the integration test for invalid cache, we already had for previously existing cache.
